### ISSUE TITLE
`getwininfo()` reports wrong bottom line when it is wrapped.

### ISSUE THREAD

**Issue by L3MON4D3** [neovim/neovim#14312]
`getwininfo()` reports wrong bottom line when it is wrapped.

- `nvim --version`: NVIM v0.5.0-dev+1226-g370469be2
- `vim -u DEFAULTS` (version: VIM - Vi IMproved 8.2) behaves differently? Neovim shows the line and appends `@@@`, while vim shows only `@@@` in place of the entire line, both report the same line number.
- Operating system/version:
- Terminal name/version:
- `$TERM`:

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
# Alternative for shell-related problems:
# env -i TERM=ansi-256color "$(which nvim)"

```

**Comment by jdhao** [comment]
Have you read the doc on `getwininfo()`, the doc says:

> 		Each List item is a Dictionary with the following entries:
>			botline		last displayed buffer line

IIUIC, `botline` is the last displayed line (line 108), not last line of the buffer. I think the output is exactly the same as what the documentation says.

**Comment by L3MON4D3** [comment]
I think the example-file is bit misleading, srry for that.
If you look at the screenshot above, `getwininfo()` returns 69+24 as `botline`, whereas, with 'last displayed buffer line', I would have expected 69+25 or 69+24 and some flag like 'last_line_wrapped' to be set, so that one can tell that 69+25 is actually visible.

**Comment by jdhao** [comment]
Then update your issue, and give your complete reproduce steps.

**Comment by L3MON4D3** [comment]
Srry, should've mentioned that the last line ofc. has to wrap.

**Comment by chrisbra** [comment]
can you please show a screenshot with simple numbering turned on? That makes it much easier to see the issue I think

**Comment by jdhao** [comment]
Can not reproduce with a file with wrap on. The botline is reported correctly. 

nvim version: latest master.

**Comment by L3MON4D3** [comment]
![screen](https://user-images.githubusercontent.com/41961280/113900135-cf7b8280-97cd-11eb-9c2b-cb5779071c79.png)
`getwininfo()` reports `botline` as 108, but line 109 is visible. I don't necessarily think that 108 is wrong, but there should be a way to tell that 109 is visible.

**Comment by chrisbra** [comment]
Yeah, it looks like botline is the last line *completely* displayed in the window.

I think we would need a doc patch like this:
```patch
diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
index 97c968c74..603780056 100644
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -6087,7 +6087,7 @@ getwininfo([{winid}])                                     *getwininfo()*
                tab pages is returned.

                Each List item is a |Dictionary| with the following entries:
-                       botline         last displayed buffer line
+                       botline         last complete displayed buffer line
                        bufnr           number of buffer in the window
                        height          window height (excluding winbar)
                        loclist         1 if showing a location list
```

ping @brammool

**Comment by L3MON4D3** [comment]
In that case, how would I be able to tell that line 109 is visible?

**Comment by chrisbra** [comment]
I suppose something like `getwininfo()[0].topline + line('w$')`

**Comment by L3MON4D3** [comment]
`line('w$')` also returns 108 :/

**Comment by chrisbra** [comment]
oh, it worked for my minimal test. Hm, indeed, sometimes one needs to subtract one and sometimes one doesn't. Hm, how about: 

    echo getwininfo()[0].topline + winheight(0) -1

**Comment by L3MON4D3** [comment]
This works only as long as there are no other wrapped lines above.
Is there any way to determine that a line is wrapped?

**Comment by brammool** [comment]
Perhaps we should add a field that indicates how many screen lines at the bottom are used for an incomplete line?

**Comment by chrisbra** [comment]
> This works only as long as there are no other wrapped lines above.
Is there any way to determine whether a line is wrapped?

Not easily. Possibly store the window view using `winsaveview(), position the cursor on the bottom line (`z-`) and return that line using line('.') and then restore the view using `winrestview()`.

Might be easier to add an additional property, mentioning the number of partial visible screen lines.

**Comment by L3MON4D3** [comment]
> Perhaps we should add a field that indicates how many screen lines at the bottom are used for an incomplete line?

In this case that would mean `botline` = 108 and `incomplete_lines_bottom` = 1?
